aten/src/ATen/core/ivalue_inl.h

```diff
@@ -508,20 +508,34 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
    */
   template <typename T>
   c10::intrusive_ptr<Future> then(T callback, TypePtr type) {
+    using IValueWithDataPtrs = std::
+        tuple<IValue, std::vector<std::reference_wrapper<const at::DataPtr>>>;
 #if __cpp_lib_is_invocable >= 201703
     static_assert(
-        std::is_invocable_r<IValue, T, Future&>::value,
-        "The callback must have signature IValue(Future&)");
+        guts::disjunction<
+            std::is_invocable_r<IValue, T, Future&>,
+            std::is_invocable_r<IValueWithDataPtrs, T, Future&>>::value,
+        "The callback must have signature IValue(Future&) or "
+        "std::tuple<IValue, std::vector<std::reference_wrapper<const DataPtr>>>(Future&)");
 #endif
     auto childFut = createInstance(std::move(type));
-    addCallback(
-        [childFut, cb = std::move(callback)](Future& parentFut) mutable {
-          try {
-            childFut->markCompleted(cb(parentFut));
-          } catch (std::exception&) {
-            childFut->setError(std::current_exception());
-          }
-        });
+    addCallback([childFut,
+                 cb = std::move(callback)](Future& parentFut) mutable {
+      try {
+        guts::if_constexpr<std::is_convertible<
```

Comment:
curious, what's the difference between `guts::if_constexpr` and `if constexpr`? Is it because we are not using C++17 yet?

Reply:
Correct. In fact, `guts::if_constexpr` will use `if constexpr` internally if available, but will fall back onto a C++-14-compatible polyfill if not.